File: smoke_tests/php/regex_dos.php
<?php
// Regular Expression DoS (ReDoS) vulnerabilities in PHP

// Test 1: Nested quantifiers
function validate_nested() {
    $input = $_GET['input'];
    // VULNERABLE: (a+)+ causes exponential backtracking
    return preg_match('/^(a+)+b$/', $input);
}

// Test 2: Overlapping alternation
function check_overlap() {
    $input = $_GET['input'];
    // VULNERABLE: Overlapping alternatives
    return preg_match('/^(a|a)+b$/', $input);
}

// Test 3: Email validation ReDoS
function validate_email() {
    $email = $_POST['email'];
    // VULNERABLE: Classic email ReDoS pattern
    $pattern = '/^([a-zA-Z0-9]+)+@([a-zA-Z0-9]+)+\.([a-zA-Z]+)+$/';
    return preg_match($pattern, $email);
}

// Test 4: URL validation ReDoS
function validate_url() {
    $url = $_GET['url'];
    // VULNERABLE: Nested groups in URL pattern
    $pattern = '/^(https?:\/\/)?([a-zA-Z0-9.-]+)+(\/.*)*$/';
    return preg_match($pattern, $url);
}

// Test 5: User-supplied regex
function custom_match() {
    $input = $_POST['input'];
    $pattern = $_POST['pattern'];
    // VULNERABLE: User supplies regex pattern
    return preg_match($pattern, $input);
}

// Test 6: preg_replace with vulnerable pattern
function replace_pattern() {
    $input = $_POST['input'];
    // VULNERABLE: Vulnerable pattern in preg_replace
    return preg_replace('/(a+)+/', 'X', $input);
}

// Test 7: preg_match_all with backtracking
function match_all() {
    $input = $_GET['input'];
    // VULNERABLE: Match all with exponential pattern
    preg_match_all('/(.+)+x/', $input, $matches);
    return count($matches[0]);
}

// Test 8: preg_split ReDoS
function split_input() {
    $input = $_POST['input'];
    // VULNERABLE: Split can trigger ReDoS
    return preg_split('/(\s+)+/', $input);
}

// Test 9: HTML tag matching ReDoS
function match_html() {
    $html = $_POST['html'];
    // VULNERABLE: Complex HTML pattern
    preg_match_all('/<([a-z]+)([^>]*)*>/', $html, $matches);
    return count($matches[0]);
}

// Test 10: Multiline ReDoS
function multiline_match() {
    $content = $_POST['content'];
    // VULNERABLE: Multiline with backtracking
    return preg_match('/^(.+)+$/m', $content);
}

// Test 11: Regex from database
function dynamic_regex() {
    $id = $_GET['id'];
    $input = $_GET['input'];
    // Assume pattern comes from database
    $pattern = get_pattern_from_db($id);
    // VULNERABLE: Pattern from database (could be user-supplied)
    return preg_match($pattern, $input);
}

// Test 12: PCRE recursive pattern
function recursive_match() {
    $input = $_GET['input'];
    // VULNERABLE: Recursive patterns can cause stack overflow
    $pattern = '/(?:a(?:a(?:a(?:a(?:a)*)*)*)*)+/';
    return preg_match($pattern, $input);
}

// Test 13: ereg (deprecated but still risky if used)
function ereg_match() {
    $input = $_GET['input'];
    // VULNERABLE: Deprecated and potentially ReDoS prone
    // return ereg('(a+)+', $input);  // Would fail in PHP 7+
}

function get_pattern_from_db($id) {
    // Placeholder
    return '/^test$/';
}
?>

File: smoke_tests/csharp/regex_dos.cs
// Regular Expression DoS (ReDoS) vulnerabilities in C#
using System;
using System.Text.RegularExpressions;
using System.Web.Mvc;

namespace VulnerableApp
{
    public class RegexController : Controller
    {
        // Test 1: Nested quantifiers
        public ActionResult ValidateNested()
        {
            string input = Request.QueryString["input"];
            // VULNERABLE: (a+)+ causes exponential backtracking
            var regex = new Regex(@"(a+)+b");
            bool match = regex.IsMatch(input);
            return Json(match);
        }

        // Test 2: Overlapping alternation
        public ActionResult ValidateOverlap()
        {
            string input = Request.QueryString["input"];
            // VULNERABLE: Overlapping alternatives
            var regex = new Regex(@"(a|a)+b");
            return Json(regex.IsMatch(input));
        }

        // Test 3: Email validation with ReDoS
        public ActionResult ValidateEmail()
        {
            string email = Request.Form["email"];
            // VULNERABLE: Classic email ReDoS pattern
            var regex = new Regex(@"^([a-zA-Z0-9]+)+@([a-zA-Z0-9]+)+\.([a-zA-Z]+)+$");
            return Json(regex.IsMatch(email));
        }

        // Test 4: URL validation ReDoS
        public ActionResult ValidateUrl()
        {
            string url = Request.QueryString["url"];
            // VULNERABLE: Nested groups in URL pattern
            var regex = new Regex(@"^(https?://)?([a-zA-Z0-9.-]+)+(/.*)*$");
            return Json(regex.IsMatch(url));
        }

        // Test 5: User-supplied regex pattern
        public ActionResult CustomMatch()
        {
            string input = Request.Form["input"];
            string pattern = Request.Form["pattern"];
            // VULNERABLE: User can supply malicious pattern
            var regex = new Regex(pattern);
            return Json(regex.IsMatch(input));
        }

        // Test 6: No timeout set
        public ActionResult NoTimeout()
        {
            string input = Request.QueryString["input"];
            // VULNERABLE: No timeout, can hang forever
            var regex = new Regex(@"(.*)*evil");
            return Json(regex.IsMatch(input));
        }

        // Test 7: Static regex with ReDoS pattern
        private static readonly Regex BadPattern = new Regex(@"([a-z]+)*$");

        public ActionResult StaticRegex()
        {
            string input = Request.QueryString["input"];
            // VULNERABLE: Static pattern with backtracking
            return Json(BadPattern.IsMatch(input));
        }

        // Test 8: Compiled regex with ReDoS
        public ActionResult CompiledRegex()
        {
            string input = Request.QueryString["input"];
            // VULNERABLE: Compiled doesn't prevent ReDoS
            var regex = new Regex(@"(x+x+)+y", RegexOptions.Compiled);
            return Json(regex.IsMatch(input));
        }

        // Test 9: Replace with ReDoS
        public ActionResult ReplacePattern()
        {
            string input = Request.Form["input"];
            // VULNERABLE: Replace also affected by ReDoS
            var regex = new Regex(@"(a+)+");
            string result = regex.Replace(input, "X");
            return Content(result);
        }

        // Test 10: Split with ReDoS
        public ActionResult SplitPattern()
        {
            string input = Request.Form["input"];
            // VULNERABLE: Split can trigger ReDoS
            var regex = new Regex(@"(\s+)+");
            string[] parts = regex.Split(input);
            return Json(parts);
        }

        // Test 11: Multiline ReDoS
        public ActionResult MultilinePattern()
        {
            string input = Request.Form["content"];
            // VULNERABLE: Multiline doesn't prevent ReDoS
            var regex = new Regex(@"^(.+)+$", RegexOptions.Multiline);
            return Json(regex.IsMatch(input));
        }

        // Test 12: HTML tag matching ReDoS
        public ActionResult MatchHtmlTag()
        {
            string html = Request.Form["html"];
            // VULNERABLE: Complex HTML pattern
            var regex = new Regex(@"<([a-z]+)([^>]*)*>");
            return Json(regex.Matches(html).Count);
        }
    }
}
